### DISCUSSION TITLE
ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: cannot allocate memory in static TLS block

### DISCUSSION THREAD

**Discussion by lssatvik** [apache/airflow#40562] [General]
[ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: cannot allocate memory in static TLS block](https://github.com/apache/airflow/issues/17546#top)

I faced this issue while performing airflow db upgrade in my setup: airflow 2.9.2, python 3.12, ubuntu 22.04, mysql-server 8.4.0
LD preload solution as described in the linked issue did solve the issue, but so did importing the package MySQLdb before anything else by adding an import line in ./py_env/bin/airflow file

![image](https://github.com/apache/airflow/assets/3456652/a4f5fef2-2ed2-46a3-8691-880f8797065e)

Placing the import line after 'import main' caused the issue again. So thinking it is a possible package issue, I did a binary search to find that 're2' package is related to the issue.

This causes the issue
![image](https://github.com/apache/airflow/assets/3456652/26054137-e0f7-42f2-90bd-b5e07ae434de)

and this doesn't
![image](https://github.com/apache/airflow/assets/3456652/74efb00e-8091-416d-b075-759ee2be8d29)

Package version google-re2==1.1.20240601 is the version installed after applying constraints. Downgrading it to google-re2==1.1 resolved the issue.

Can I assume it is this package that's causing the issue? Or is it due to some buggy library? I did install many libraries to build mysqlclient wheel (libmysqlclient-dev python3-dev libmysqlclient21 default-libmysqlclient-dev build-essential gcc)

Funny thing is, when i recreated this environment on ubuntu 20.04 with mysql-server 8.0.37, this libstdc++ issue never occurred even with google-re2==1.1.20240601 installed. Does this mean mysql 8.4.0 is the culprit?

Also, the docs say mysql 8.0 and innovation are supported versions, does 8.4.0 LTS not come under this? I didn't totally understand what they said in the oracle link in the docs about innovation. They say 8.x is all innovation.

google re 1.1.20240601 is an equality constraint. Can I use 1.1 here? Or is it better to simply go for hackish import way. Or should i downgrade mysql?

Summary picture:-
![image](https://github.com/apache/airflow/assets/3456652/52c01f7d-3b8c-4bf9-a09f-ffa9ffb41c03)

_Originally posted by @lssatvik in https://github.com/apache/airflow/issues/17546#issuecomment-2198324967_

**Comment by potiuk** [comment]
Likely bad combination of packages. And combination of bugs in c-code of mysql - it's known to cause issues and LD_PRELOAD is the recommended solution. I think loosing time on analysing root cause of c-libraries clashing with each other is genrally a wasted effort.

You can raise the issue to mysql - BTW . We are considering dropping MysQL as supported backend in Airflow 3 because od a number of problems like this - so I recommend to use Postgres.

**Reply by lssatvik** [reply]
Alright, thank you.

**Reply by lucasfcnunes** [reply]
@potiuk does it make sense to add

```Dockerfile
ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libstdc++.so.6
```

to Airflow (this repo) images/builds to solve it for everybody?

**Reply by potiuk** [reply]
No. This causes that every binary started also preloads this library (which is big) - so it adds a lot of overhead. This is a workaround for buggy mysql packages - and we certainly do not want to slow performance of everyone in order to workaround this issue.
